system/jlib/jsocket.cpp

```diff
@@ -5081,51 +5093,89 @@ class CSocketEpollThread: public CSocketBaseThread
 
 class CSocketEpollHandler: implements ISocketSelectHandler, public CInterface
 {
-    CSocketEpollThread *epollthread;
+    CIArrayOf<CSocketEpollThread> threads;
     CriticalSection sect;
+    bool started;
     StringAttr epolltrace;
 public:
     IMPLEMENT_IINTERFACE;
     CSocketEpollHandler(const char *trc)
         : epolltrace(trc)
     {
-        epollthread = new CSocketEpollThread(epolltrace);
+        started = false;
     }
 
     ~CSocketEpollHandler()
     {
-        delete epollthread;
+        stop(true);
+        threads.kill();
     }
 
     void start()
     {
         CriticalBlock block(sect);
-        epollthread->start();
+        if (!started)
+        {
+            started = true;
+            ForEachItemIn(i,threads)
+            {
+                threads.item(i).start();
+            }
+        }
     }
 
     void add(ISocket *sock,unsigned mode,ISocketSelectNotify *nfy)
     {
-        CriticalBlock block(sect); // JCS->MK - are these blocks necessary? epollthread->add() uses it's own CS.
-
-        /* JCS->MK, the CSocketSelectHandler variety, checks result of thread->add and spins up another handler
-         * Shouldn't epoll version do the same?
-         */
-        if (!epollthread->add(sock,mode,nfy))
-            throw MakeStringException(-1, "CSocketEpollHandler: failed to add socket to epollthread handler: sock # = %d", sock->OShandle());
+        CriticalBlock block(sect);
+        // Create new handler thread if current one has many fds
+        // epoll() handles many fds faster than select so this would
+        // seem not as important, but we are still serializing on
+        // nfy events and spreading those over a few threads may help
+        for (;;)
+        {
+            bool added=false;
+            ForEachItemIn(i,threads)
+            {
+                if (added)
+                    threads.item(i).remove(sock);
```

Comment:
if remove returns true, then there can't be any other occurrences (in other threads), so could break out here.

Reply:
Yes, good opt, I will add this.

Reply:
changed